### ISSUE TITLE
How to grant removableStorage access?

### ISSUE THREAD

**Issue by forderud** [microsoft/SandboxSecurityTools#6]
Sorry if submitting this request to the incorrect audience, but I'm struggling to get in contact with someone who's knowledgeable in this area.

Can someone please help me figure out how to grant AppContainer access to USB removable storage that's formatted with a permission-free filesystem like FAT32?

### Steps to reproduce
* Insert a USB stick with a FAT32 filesystem
* Start `LaunchAppContainer.exe -m MyAppContainer -i notepad.exe -c S-1-15-3-10` (SID for "removableStorage" capability mentioned on https://learn.microsoft.com/en-us/windows/uwp/packaging/app-capability-declarations)
* From [Process Explorer](https://learn.microsoft.com/en-us/sysinternals/downloads/process-explorer), observe that notepad starts up with "Removable storage" capability:
![image](https://user-images.githubusercontent.com/2671400/208636491-acc50a24-193a-42de-acfd-28a2f631a8ff.png)
* From notepad, press File -> Open and attempt to access the USB stick.
* Observe the following "Access is denied" error:
![image](https://user-images.githubusercontent.com/2671400/208636392-fba184eb-70b4-4698-930e-bacfc19a2bb6.png)

**Comment by forderud** [comment]
Gentle reminder.

**Comment by rmicelims** [comment]
The removableStorage capability is designed to be used with APIs from the Windows Runtime, such as the Windows.Storage namespace, it has no effect on classic win32 file APIs. Even though the file system in question does not support permissions there is still a security descriptor applied to the storage device hosting the volumes. In the case of USB storage volumes it's most likely using the default security descriptor provided by the IO manager, which does not allow for app container traversal into the volume.

**Comment by forderud** [comment]
> In the case of USB storage volumes it's most likely using the default security descriptor provided by the IO manager, which does not allow for app container traversal into the volume.

Thanks for the explanation @rmicelims . Is there then any way of _modifying_ the default security descriptor you're referring to, so that it becomes accessible from an app container? I'm thinking machine-wide security policy settings or similar.

**Comment by rmicelims** [comment]
I don't believe so. The security descriptor is hard coded into the IO manager/associated drivers in the IO stack.

Upon further investigation it seems like the removableStorage capability was intended to be used with Windows Phone or other portable devices, with limited support for Desktop SKUs: https://learn.microsoft.com/en-us/windows/uwp/files/access-the-sd-card

I believe you will need to use the various file picker classes in order to access removable storage devices: https://learn.microsoft.com/en-us/windows/uwp/files/quickstart-using-file-and-folder-pickers

However, this requires setting some properties in an app package manifest, and won't work if you're simply trying to run an existing Win32 app as AppContainer.

**Comment by rmicelims** [comment]
I would recommend checking out the new App Isolation feature that's in public preview. It's intended to be used in scenarios where developers want to run an existing Win32 application in AppContainer, with additional support for things like file accesses: https://blogs.windows.com/windowsdeveloper/2023/06/14/public-preview-improve-win32-app-security-via-app-isolation/

**Comment by forderud** [comment]
> I would recommend checking out the new App Isolation feature that's in public preview. It's intended to be used in scenarios where developers want to run an existing Win32 application in AppContainer, with additional support for things like file accesses: https://blogs.windows.com/windowsdeveloper/2023/06/14/public-preview-improve-win32-app-security-via-app-isolation/

My Win32 "applications" in question are actually COM servers in a plugin architecture. They are already compatible with being launched in an AppContainer process through `CreateProcessAsUser` with `PROC_THREAD_ATTRIBUTE_SECURITY_CAPABILITIES` parameters - just as in the LaunchAppContainer tool. Everything then works fine, except for USB FAT32 drive access, that's being blocked by the AppContainer.

Can the new App Isolation feature still be used for my scenario where I rely on programmatic process creation?

**Comment by rmicelims** [comment]
App Isolation will work for programmatic process creation, however, I believe there will be some UI interaction required in the form of a consent dialog to get access to files on a USB device.
